### PR TITLE
Add nested squeeze.dim and unsqueeze

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #86813

### CODE REVIEW

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 1]
```diff
@@ -1029,6 +1029,59 @@ Tensor transpose_nested(const Tensor& self, int64_t dim0, int64_t dim1) {
...
+  auto self_ptr = get_nested_tensor_impl(self);
+  int64_t ndim = self_ptr->dim();
+  int64_t wrapped_dim = at::maybe_wrap_dim(dim, ndim);
+  TORCH_CHECK(wrapped_dim > 0,
+  "squeeze(): For nested tensors, squeezing dimension 0 is not supported at the moment");
+  const Tensor& sizemat = self_ptr->get_nested_size_tensor();
+  const Tensor& stridemat = self_ptr->get_nested_stride_tensor();
+  // if tensor.size(dim) != 1 torch.squeeze will return the result, we do the same here
+  auto size_dim = sizemat.select(1, wrapped_dim - 1);
+  if (!at::all(size_dim == 1).item<bool>()) {
```
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:1042]: could probably use opt_sizes here

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 2]
```diff
@@ -1029,6 +1029,59 @@ Tensor transpose_nested(const Tensor& self, int64_t dim0, int64_t dim1) {
       self, sizemat_transposed, stridemat_transposed, std::vector<int64_t>(self_ptr->get_storage_offsets()));
 }
 
+Tensor squeeze_nested(const Tensor& self, int64_t dim) {
+  auto self_ptr = get_nested_tensor_impl(self);
+  int64_t ndim = self_ptr->dim();
+  int64_t wrapped_dim = at::maybe_wrap_dim(dim, ndim);
+  TORCH_CHECK(wrapped_dim > 0,
+  "squeeze(): For nested tensors, squeezing dimension 0 is not supported at the moment");
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:1037]: nit: Ask the user to open a GH issue if and why they need this.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 3]
```diff
@@ -1029,6 +1029,59 @@ Tensor transpose_nested(const Tensor& self, int64_t dim0, int64_t dim1) {
...
+  return create_nested_view_tensor(
+      self, sizemat_squeezed, stridemat_squeezed, std::vector<int64_t>(self_ptr->get_storage_offsets()));
+}
+
+Tensor unsqueeze_nested(const Tensor& self, int64_t dim) {
+  auto self_ptr = get_nested_tensor_impl(self);
+  int64_t ndim = self_ptr->dim();
+  TORCH_CHECK(dim >= - ndim - 1 && dim <= ndim,
+  "unsqueeze(): dim must be in [-self.dim() - 1, self.dim() + 1), got", dim);
+  int64_t wrapped_dim = dim < 0 ? dim + ndim + 1 : dim;
```
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:1061]: There is probably a reason why at::maybe_wrap_dim doesn't work here right?

File: test/test_nestedtensor.py [hunk 4]
```diff
@@ -1762,6 +1790,30 @@ def test_nested_tensor_reshape_backward(self):
...
+        nt = torch.nested.nested_tensor([torch.randn((2, 6)), torch.randn((3, 6))], requires_grad=True)
+        with torch.no_grad():
+            pt = torch.nested.to_padded_tensor(nt, 0.0).requires_grad_(True)
+
+        ynt = nt.unsqueeze(2)
+        ypt = pt.unsqueeze(2)
+        ynt.backward(ynt.clone())
+        ypt.backward(ypt.clone())
+
+        self.assertEqual(torch.nested.to_padded_tensor(nt.grad, 0.0), pt.grad)
```
Comment (drisspg) [inline test/test_nestedtensor.py:1815]: Curious if you tried testing this via the gradcheck way of things?

Comment (drisspg) [reply test/test_nestedtensor.py:1815]: This looks good though

**Review by drisspg — APPROVED** [review]
Small things, but looks good

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 3 above)
Comment (mikaylagawarecki) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:1061]: oh ur right this is just `wrapped_dim = maybe_wrap_dim(dim, self.dim() + 1);` thanks

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 5]
```diff
@@ -1029,6 +1029,59 @@ Tensor transpose_nested(const Tensor& self, int64_t dim0, int64_t dim1) {
...
+  "squeeze(): For nested tensors, squeezing dimension 0 is not supported at the moment");
+  const Tensor& sizemat = self_ptr->get_nested_size_tensor();
+  const Tensor& stridemat = self_ptr->get_nested_stride_tensor();
+  // if tensor.size(dim) != 1 torch.squeeze will return the result, we do the same here
+  auto size_dim = sizemat.select(1, wrapped_dim - 1);
+  if (!at::all(size_dim == 1).item<bool>()) {
+    // detach to avoid triggering throw_error_if_base_and_tensor_are_same
+    return self.detach();
+  }
+  auto column_indices = sizemat.new_empty(ndim - 2);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:1046]: Does this function work with ndim is 2?

File: test/test_nestedtensor.py [hunk 6]
```diff
@@ -1290,6 +1290,34 @@ def test_transpose(self, device, dtype):
         ptT = pt.transpose(-1, -2)
         self.assertEqual(ptT, ptT_from_ntT)
 
+    @dtypes(torch.float, torch.float16, torch.double)
+    def test_squeeze_unsqueeze(self, device, dtype):
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:1294]: Do we currently get failure when we try to call `squeeze` without providing a dim or is `squeeze` implemented as a loop of `squeeze.dim` calls?

Comment (mikaylagawarecki) [reply test/test_nestedtensor.py:1294]: it will fail, `func: squeeze` is registered separately from `squeeze.dim` https://www.internalfb.com/code/fbsource/[02f985b971a253089937b0a5347497087bb72fd4]/fbcode/caffe2/aten/src/ATen/native/native_functions.yaml?lines=4470

Comment (cpuhrsch) [reply test/test_nestedtensor.py:1294]: We could add a stub that throws a more meaningful error message and refers users to use `squeeze` with a dim argument directly. You can register a function that does nothing more than throw an exception and ask users to open an issue if and why they need it.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp (hunk 5 above)
Comment (mikaylagawarecki) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:1046]: Hm, this case returns a tensor iiuc, so we just error out?

Comment (cpuhrsch) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:1046]: It's ok to throw an exception for this PR, take note in a separate issue and follow up

Comment (mikaylagawarecki) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:1046]: In general, do nested tensors support having scalars as the nested elements?
